File: internal/dashboard/subagent_status_test.go
package dashboard

import (
	"testing"

	agent "github.com/gratefulagents/sdk/pkg/agentsdk"
)

func TestClassifySubagentStatus(t *testing.T) {
	cases := map[string]subagentStatusCategory{
		"":                subagentStatusLive,
		"running":         subagentStatusLive,
		"started":         subagentStatusLive,
		"Running":         subagentStatusLive,
		" completed ":     subagentStatusSucceeded,
		"succeeded":       subagentStatusSucceeded,
		"failed":          subagentStatusFailed,
		"timeout":         subagentStatusFailed,
		"timed_out":       subagentStatusFailed,
		"error":           subagentStatusFailed,
		"killed":          subagentStatusFailed,
		"stopped":         subagentStatusStopped,
		"cancelled":       subagentStatusStopped,
		"canceled":        subagentStatusStopped,
		"waiting":         subagentStatusWaiting,
		"pending":         subagentStatusWaiting,
		"reconciling":     subagentStatusWaiting,
		"dependency_wait": subagentStatusWaiting,
		"something_new":   subagentStatusUnknown,
		"ok":              subagentStatusUnknown,
	}
	for status, want := range cases {
		if got := classifySubagentStatus(status); got != want {
			t.Errorf("classifySubagentStatus(%q) = %q, want %q", status, got, want)
		}
	}
}

func TestIsTerminalSubagentStatusUsesCategories(t *testing.T) {
	for _, status := range []string{"completed", "succeeded", "failed", "timeout", "error", "stopped", "cancelled", "canceled"} {
		if !isTerminalSubagentStatus(status) {
			t.Errorf("%q should be terminal", status)
		}
	}
	for _, status := range []string{"", "running", "started", "waiting", "pending", "reconciling", "something_new"} {
		if isTerminalSubagentStatus(status) {
			t.Errorf("%q must not be terminal", status)
		}
	}
}

// A status the SDK has never emitted before must not leave the task alive
// forever: any terminal category closes it, unknown strings stay progress.
func TestContentEventToActivityEntryRoutesTerminalStatusesToCompleted(t *testing.T) {
	cases := map[string]string{
		"started":       "subagent_started",
		"completed":     "subagent_completed",
		"failed":        "subagent_completed",
		"cancelled":     "subagent_completed",
		"stopped":       "subagent_completed",
		"timeout":       "subagent_completed",
		"error":         "subagent_completed",
		"running":       "subagent_progress",
		"waiting":       "subagent_progress",
		"something_new": "subagent_progress",
	}
	for status, wantType := range cases {
		entry := contentEventToActivityEntry(&agent.ContentEvent{
			Type:      "subagent_status",
			Status:    status,
			AgentName: "reviewer",
			Message:   "review the diff",
		})
		if entry == nil {
			t.Fatalf("status %q produced no entry", status)
		}
		if entry.Type != wantType {
			t.Errorf("status %q routed to %q, want %q", status, entry.Type, wantType)
		}
		if entry.SubagentStatus != status {
			t.Errorf("status %q not preserved on entry: %q", status, entry.SubagentStatus)
		}
	}
}

File: platform-app/frontend/src/lib/subagentStatus.test.ts
import { describe, expect, it } from "vitest";
import {
  classifySubagentStatus,
  isLiveSubagentStatus,
  isTerminalSubagentStatus,
  isWaitingSubagentStatus,
  subagentStatusLabel,
} from "./subagentStatus";

describe("classifySubagentStatus", () => {
  it("buckets the SDK vocabulary", () => {
    expect(classifySubagentStatus("")).toBe("live");
    expect(classifySubagentStatus("running")).toBe("live");
    expect(classifySubagentStatus("started")).toBe("live");
    expect(classifySubagentStatus("waiting")).toBe("waiting");
    expect(classifySubagentStatus("pending")).toBe("waiting");
    expect(classifySubagentStatus("reconciling")).toBe("waiting");
    expect(classifySubagentStatus("completed")).toBe("succeeded");
    expect(classifySubagentStatus("succeeded")).toBe("succeeded");
    expect(classifySubagentStatus("failed")).toBe("failed");
    expect(classifySubagentStatus("stopped")).toBe("stopped");
    expect(classifySubagentStatus("cancelled")).toBe("stopped");
    expect(classifySubagentStatus("canceled")).toBe("stopped");
  });

  it("treats timeout and error spellings as failures, not success", () => {
    for (const status of ["timeout", "timed_out", "error", "errored", "killed"]) {
      expect(classifySubagentStatus(status)).toBe("failed");
      expect(isTerminalSubagentStatus(status)).toBe(true);
    }
  });

  it("is case- and whitespace-insensitive", () => {
    expect(classifySubagentStatus(" Completed ")).toBe("succeeded");
    expect(classifySubagentStatus("FAILED")).toBe("failed");
  });

  it("fails closed on unknown strings", () => {
    expect(classifySubagentStatus("something_new")).toBe("unknown");
    expect(classifySubagentStatus("constructor")).toBe("unknown");
    expect(isTerminalSubagentStatus("something_new")).toBe(false);
    expect(isLiveSubagentStatus("something_new")).toBe(false);
    expect(isWaitingSubagentStatus("something_new")).toBe(false);
    expect(classifySubagentStatus(undefined)).toBe("live");
    expect(classifySubagentStatus(null)).toBe("live");
  });

  it("labels non-live statuses for badges", () => {
    expect(subagentStatusLabel("failed")).toBe("failed");
    expect(subagentStatusLabel("timeout")).toBe("failed (timeout)");
    expect(subagentStatusLabel("cancelled")).toBe("stopped");
    expect(subagentStatusLabel("waiting")).toBe("waiting");
    expect(subagentStatusLabel("something_new")).toBe("unknown status: something_new");
  });
});
